exp/api/v1alpha4/azuremanagedmachinepool_webhook.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"context"
+	"errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var azuremanagedmachinepoollog = logf.Log.WithName("azuremanagedmachinepool-resource")
+
+var webhookClient client.Client
+
+// SetupWebhookWithManager sets up and registers the webhook with the manager.
+func (r *AzureManagedMachinePool) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// NewAzureManagedMachinePoolWebhook returns a new AzureManagedMachinePool instance.
+func NewAzureManagedMachinePoolWebhook(client client.Client) *AzureManagedMachinePool {
+	webhookClient = client
```

Comment:
please do not make this a package level variable set inside a constructor.

I think you should use the lower level webhook implementation like in https://github.com/alexeldeib/hook-test/blob/master/webhook.go

sorry I missed this on earlier reviews

Reply:
you might be able to get away with wrapping the CRD type with a "fake" type which implements CRD + Client and implements Defaulter, but I'm not sure controller-runtime will be happy with that.

Reply:
yeah looks like no dice, CR will check if the object it got the webhook request for implements validator/defaulter: https://github.com/kubernetes-sigs/controller-runtime/blob/8b55f85c90c3b1df1af58ddb7fe50096fdc7aa99/pkg/webhook/admission/validator.go#L64

Reply:
okay sure let me change this.